### ISSUE TITLE
Introduce Snapshot Testing Like Jest

### ISSUE THREAD

**Issue by RehanSaeed** [xunit/xunit#1727]
The [Jest](https://facebook.github.io/jest/) JavaScript test framework has a feature called [snapshot testing](https://facebook.github.io/jest/docs/en/snapshot-testing.html#snapshot-testing-with-jest). It would be cool for xUnit to add this feature, especially as Blazor components are coming and snapshot testing is great for testing renderred HTML output.

**Comment by theramis** [comment]
I recently made a library for exactly this purpose. 
You can find it on Nuget, its called Snapper. The source is here https://github.com/theramis/Snapper.

**Comment by BrunoJuchli** [comment]
@RehanSaeed 
What is the benefit of this being integrated into (tightly coupled with) xunit in comparison to using an independent library like the what @theramis mentioned?

**Comment by bartelink** [comment]
this is also referred to as approval tests. [Random impl I hit when I googled it](https://github.com/approvals/ApprovalTests.Net)

**Comment by theramis** [comment]
> What is the benefit of this being integrated into (tightly coupled with) xunit in comparison to using an independent library like the what @theramis mentioned?

There are two main downsides of the library I made which could be solved integrating snapshots into xunit. 
1. The first is that updating a snapshot is cumbersome. 
2. The UX of how a diff is shown isn't great

By having snapshotting integrated into xunit you can improve the user experience of managing snapshots greatly. The xunit runner could take in parameters to update snapshots like jest does. The xunit VS plugin could provide an easy way to update and manage snapshots.

**Comment by tmort93** [comment]
I am not speaking against this proposal, merely offering additional information for consideration... 
I prefer [Snapshooter](https://swisslife-oss.github.io/snapshooter/) for snapshot testing with xUnit.

> The first is that updating a snapshot is cumbersome.

Snapshooter:
When a mismatch occurs a mismatch file is automatically generated next to the snap file and the test fails; this is default behavior. I use my diff tool of choice to compare the files and if the change is acceptable, I overwrite the snap with the mismatch, else I submit a bug. Mismatch files are automatically cleaned up when a mismatch is no longer occurring. An environment variable is available to disable mismatch file generation in build/release pipeline, deployed, or other scenarios. This approach has not been cumbersome for me.

> The UX of how a diff is shown isn't great

By using a diff tool of my choice I get the experience I want.

> The xunit runner could take in parameters to update snapshots like jest does.

I'm not against this, it just doesn't make sense for how _I_ maintain my snapshot tests, which is in Visual Studio with Test Explorer. The only time my snapshot tests get run from the command line is when they are already deployed somewhere (or when I want to test that they will run correctly from the command line). I do not (and probably would not since I've already been there with Jest) maintain snapshot files from the command line.

**I recommend the Snapshooter design over the Jest design for updating/managing snapshots.**

I've used Jest snapshot testing extensively in the past, with a full suite of integration tests, and while Snapshooter is very much based on Jest snapshot testing, I prefer the Snapshooter design when it comes to updating/managing snapshots.

Jest update process:
1. Run tests
2. If there are failures then:
 a. run tests again with update flag to update all snaps, then 'undo' changes for any snaps you don't want updated (don't miss any or you now have a false positive test!) before you add to source control
b. analyze all failures, identify all snaps that need updated, and put together a (potentially) long and complex testNamePattern in addition to the update flag so you only run tests again for tests with snaps you wish to update

Snapshooter update process:
1. Run tests
2. Navigate failures one by one and decide to update the snap with a simple copy+paste or to leave it failing until a bug is fixed
3. Cleaning up mismatch files:
 a. If you want Snapshooter to do it for you then simply run tests again one at a time, all at once, or in any combination you want
 b. Simply delete them yourself or 'undo' them in source control (you are not undoing changed snaps that could potentially introduce false positives in your tests if accidentally committed, but only the mismatch files which were never a part of source control to begin with)

The snapshooter process may _sound_ like more work but in practice I have very much preferred it.
I don't like being _forced_ to run tests again in order to update snap files (although I do tend to do that when I want Snapshooter to clean up mismatch files I no longer need). I feel like the Snapshooter design allows me more/easier control over which snaps I update and when/how I update them. When approaching test and snap file maintenance from the perspective of an IDE, the Snapshooter design seems simpler and makes more sense to me.

**Complaints I have that I would like to see resolved in an xUnit implementation**

Two of my biggest complaints with Snapshooter:
1. No (or terrible) strong type support for ignoring fields in a snap comparison
2. No easy/simple way to:
 a. Copy all snaps to the output directory on build so they can be deployed alongside your test binary
b. Tell Snapshooter to look for snaps in the output directory in a Release (deployed) scenario instead of where they normally are (by default they are generated in a _snapshot_ folder next to your test class file)

In summary: I support this proposal, but I do not think matching Jest design _exactly_ is the best approach. I think [Snapshooter](https://swisslife-oss.github.io/snapshooter/) is a great reference point and would like to see the Snapshooter design for updating/maintaining snap files persued instead of the Jest design. Finally, I would like to see some additional features/niceties in an xUnit implementation that are lacking in Snapshooter.

If any of this didn't make sense please ask questions and I'll try to clarify.
If this is totally different from how you do, or would do, snapshot testing please speak up as I'm always open to a better way, thanks!

**Comment by SimonCropp** [comment]
try https://github.com/VerifyTests/Verify 

>The first is that updating a snapshot is cumbersome.

Verify has a variety of approaches to [managing snapshots](https://github.com/VerifyTests/Verify#snapshot-management)

>No (or terrible) strong type support for ignoring fields in a snap comparison

Verify can ignore a type, a member by name, a member expression,   https://github.com/VerifyTests/Verify/blob/main/docs/serializer-settings.md#ignoring-a-type

> The UX of how a diff is shown isn't great 

Verify support all the common diff tools https://github.com/VerifyTests/DiffEngine#supported-tools

> No easy way to Tell Snapshooter to look for snaps in the output directory

Verify give full control over this https://github.com/VerifyTests/Verify/blob/main/docs/naming.md#derivepathinfo

**Comment by RehanSaeed** [comment]
I raised this some time ago before I was aware of Verify which I think solves this issue albeit using a second NuGet package.

**Comment by SimonCropp** [comment]
IMO this does not need to be included in Xunit. so probably remove it from the v3 release

**Comment by SimonCropp** [comment]
@RehanSaeed in that case should this be closed?

**Comment by SimonCropp** [comment]
@bradwilson can u remove the v3 label
